### PR TITLE
Speculative sampling

### PR DESCRIPTION
**Pull request by haim-barad**
Works with DollyV2 for machines with at least 64GB local memory. By default, we use GPT2 for smaller machines.

Leveraging speculative sampling with KV caching, this notebook's code will generate text using standard autoregressive sampling and speculative sampling and compare the times needed to generate N tokens. By default, N is set to 100. Another parameter K is set to 5 but can be adjusted to determine the number of candidate tokens to generate from the smaller draft model.

This code uses gradio to provide an interface labeling the models used and handling the input.

### CODE REVIEW

**Comment by andrei-kochin** [comment]
@haim-barad please check the CI failures as it seems some of the are quite easy to fix

File: notebooks/264-speculative-sampling/264-speculative-sampling.ipynb [hunk 1]
```diff
@@ -0,0 +1,492 @@
+{
```
Comment (apaniukov) [inline notebooks/264-speculative-sampling/264-speculative-sampling.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #31.            res = target_model(x_draft_target_input, attention_mask=torch.ones(x_draft_target_input.size(), dtype=torch.long), use_cache=False)
</pre><p>Why use_cache=False? You should use target_past_kv for target_model inference based on <code>x_draft_target_input</code>: </p><pre class="ql-syntax" spellcheck="false">if target_past_kv is None:
    x_draft_target_input = torch.cat((x, x_draft), dim=1)
else:
    x_draft_target_input = x_draft
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1410/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (apaniukov) [inline notebooks/264-speculative-sampling/264-speculative-sampling.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #37.                q_cum = torch.cat((q_cum, q), dim=1)
</pre><p><code>q_cum</code> is not used anywhere.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1410/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (apaniukov) [inline notebooks/264-speculative-sampling/264-speculative-sampling.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #46.                if np.random.random() &lt; min(1, (q_item / p_item)):  # accepted
</pre><p><code>q_item</code> is logits, <code>p_item</code> is probability after softmax. You should apply softmax to <code>q</code> first.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1410/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (apaniukov) [inline notebooks/264-speculative-sampling/264-speculative-sampling.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #58.            target_past_kv = target_new_past_kv
</pre><p>You should update <code>target_past_kv</code> based on the number of accepted tokens.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/1410/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

**Comment by haim-barad** [comment]
Code_check failure is another notebook. Spelling errors are ignoring the <spell> tag (as noted in your instructions).

**Comment by eaidova** [comment]
@haim-barad, for spell check you need to add unknown words reported by tools into vocabulary. You can find more info about that [contributing guide](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/CONTRIBUTING.md#pyspelling)

**Comment by haim-barad** [comment]
I had added the required entries and now I see the latest push created some conflicts, I apologize for this, but can someone with permissions do a merge?

**Comment by haim-barad** [comment]
code check is failing for 263-latent-consistency-models-image-generation, not this notebook.

**Comment by haim-barad** [comment]
Both code_check and docker_treon are failing on other notebooks, not because of this PR. Please approve. Thanks.

**Comment by haim-barad** [comment]
There's still a small link to fix. I see this was renumbered to 266. There's a "open in Collab" button in the readme - the link needs to be 266 instead of 265. Can one of you fix it to avoid the review process?
